api/nodes/serializers.py

```diff
@@ -130,6 +131,128 @@ def update(self, instance, validated_data):
+                                 help_text='Highest permission the user has.  Blank input defaults write permission if '
+                                           'user is being added and to current permission if user is being edited.')
+    links = LinksField({
+        'html': 'absolute_url',
+        'nodes': {
+            'relation': Link('users:user-nodes', kwargs={'user_id': '<pk>'})
+        },
+        'detail': Link('nodes:node-contributor-detail',
+                       kwargs={'node_id': '<node_id>', 'user_id': '<pk>'})
+    })
+
+    class Meta:
+        type_ = 'users'
+
+    def absolute_url(self, obj):
+        return obj.absolute_url
+
+    def create(self, validated_data):
+        auth = Auth(self.context['request'].user)
+        node = self.context['view'].get_node()
+        user = get_object_or_404(User, validated_data['_id'])
+        # Node object checks for contributor existence but can still change permissions anyway
+        if user in node.contributors:
+            raise exceptions.ValidationError('{} is already a validated contributor'.format(user.username))
+
+        bibliographic = validated_data['bibliographic']
+        permissions = self.get_permissions_list(validated_data['permission']) \
+            if 'permission' in validated_data else ['read', 'write']
+        node.add_contributor(contributor=user, auth=auth, visible=bibliographic, permissions=permissions, save=True)
+        node.reload()
+        user.permission = node.get_permissions(user)[-1]
+        user.bibliographic = node.get_visible(user)
+        user.node_id = node._id
+        return user
+
+    @staticmethod
+    def get_permissions_list(permission):
+        if permission == 'admin':
+            return ['read', 'write', 'admin']
+        elif permission == 'write':
+            return ['read', 'write']
+        elif permission == 'read':
+            return ['read']
+        else:
+            return ['read', 'write']
+
+
+class NodeContributorDetailSerializer(NodeContributorsSerializer):
+    """ Overrides node contributor serializer to make id read only and add addtional methods
+    """
+    id = ser.CharField(read_only=True, source='_id')
+
+    def update(self, instance, validated_data):
+        user = instance
+        auth = Auth(self.context['request'].user)
+        node = self.context['view'].get_node()
+        if 'permission' in validated_data and validated_data['permission'] != '':
+            self.set_user_permissions(node, validated_data['permission'], user, auth)
+        try:
+            node.set_visible(user, validated_data['bibliographic'], save=True)
+        except ValueError as e:
+            raise exceptions.ValidationError(e)
+        user.permission = node.get_permissions(user)[-1]
+        user.bibliographic = node.get_visible(user)
+        user.node_id = node._id
+        return user
+
+    # checks to make sure unique admin is removing own admin privilege
```

Comment:
The whole remove method is kind of a mess.  Rather than raise exceptions for last visible contributor or last admin it just returns false and doesn't save.  It makes it impossible to distinguish the two without an extra test and makes it require custom exception handling

```py
    def remove_contributor(self, contributor, auth, log=True):
        """Remove a contributor from this node.

        :param contributor: User object, the contributor to be removed
        :param auth: All the auth information including user, API key.
        """
        # remove unclaimed record if necessary
        if self._primary_key in contributor.unclaimed_records:
            del contributor.unclaimed_records[self._primary_key]

        self.contributors.remove(contributor._id)

        self.clear_permission(contributor)
        if contributor._id in self.visible_contributor_ids:
            self.visible_contributor_ids.remove(contributor._id)

        if not self.visible_contributor_ids:
            return False

        # Node must have at least one registered admin user
        # TODO: Move to validator or helper
        admins = [
            user for user in self.contributors
            if self.has_permission(user, 'admin')
            and user.is_registered
        ]
        if not admins:
            return False

        # Clear permissions for removed user
        self.permissions.pop(contributor._id, None)

        # After remove callback
        for addon in self.get_addons():
            message = addon.after_remove_contributor(self, contributor, auth)
            if message:
                status.push_status_message(message, kind='info', trust=True)

        if log:
            self.add_log(
                action=NodeLog.CONTRIB_REMOVED,
                params={
                    'project': self.parent_id,
                    'node': self._primary_key,
                    'contributor': contributor._id,
                },
                auth=auth,
                save=False,
            )

        self.save()
```